cpp/inc/bond/ext/grpc/wait_callback.h

```diff
@@ -0,0 +1,175 @@
+    /// the first one that is invoked will be recorded.
+    ///
+    /// It is safe for callbacks to outlive the wait_callback that created
+    /// them. When they are invoked their results will be discarded.
+    CallbackType callback()
+    {
+        // Create a weak_ptr to the implementation for the callback. If all
+        // the wait_callback instances sharing the implementation go away,
+        // their callbacks can still safely be invoked: they just won't be
+        // able to lock the weak_ptr and thus won't do anything.
+        std::weak_ptr<impl> wpImpl(_impl);
+
+        return std::function<void(const bond::bonded<TResponse>&, const grpc::Status&)>(
+            [wpImpl](const bond::bonded<TResponse>& response, const grpc::Status& status)
+            {
+                std::shared_ptr<impl> impl = wpImpl.lock();
+
+                if (impl)
+                {
+                    impl->set(response, status);
+                }
+            });
+    }
+
+    /// @brief Implicitly creates a callback.
+    operator CallbackType()
+    {
+        return callback();
+    }
+
+    /// @brief Waits for a callback created by \ref callback to be
+    /// invoked.
+    void wait() const
+    {
+        _impl->wait();
+    }
+
+    /// @brief Waits at least \p timeout for a callback created by \ref
+    /// callback to be invoked.
+    ///
+    /// @param timeout the minimum amount of time to wait.
+    ///
+    /// @return \p true if a callback was invoked. \p false if the timeout
+    /// occured.
+    template <typename Rep, typename Period>
+    bool wait(std::chrono::duration<Rep, Period> timeout) const
+    {
+        return _impl->wait(timeout);
+    }
+
+    /// @brief Gets the response.
+    ///
+    /// @warning Blocks until a callback is invoked.
+    const bond::bonded<TResponse>& response() const
+    {
+        return _impl->response();
+    }
+
+    /// @brief Gets the status.
+    ///
+    /// @warning Blocks until a callback is invoked.
+    const grpc::Status& status() const
+    {
+        return _impl->status();
+    }
+
+private:
+    /// The interesting guts of wait_callback. We use an impl class so that
+    /// we can handout weak_ptr instances to be captured by the callbacks
+    /// without forcing clients of wait_callback to deal with creating a
+    /// shared_ptr.
+    struct impl
+    {
+        /// mutex to lock the shared state
+        mutable std::mutex _m;
+        /// condition variable used to signal anyone waiting
+        mutable std::condition_variable _cv;
+        /// The response, but more importantly, doubles as a flag indicating
+        /// whether a callback has been invoked yet. If this is empty, no
+        /// callback has been invoked yet. If non-empty, a callback has
+        /// already been invoked.
+        boost::optional<bond::bonded<TResponse>> _response;
+        /// The status.
+        boost::optional<grpc::Status> _status;
```

Comment:
Do we really need two separate `boost::optional`-s instead of a single one for the aggregate type (they seem to be updated together)?